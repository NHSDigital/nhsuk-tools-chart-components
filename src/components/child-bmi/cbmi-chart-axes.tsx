import { ChildBmiResultChart } from '@models/cbmi-chart.model';
import { BaseClassifications } from '@custom-types/bmi.types';
import React, { JSX } from 'react';
import './styles/cbmi-chart-axes.module.scss';

type ChildBmitGraphAxisMarkerProps = {
  markerText: string;
  offsetPercent: number;
};

type ChildBmitGraphAxesProps = {
  model: ChildBmiResultChart;
};

function AxisMarker({
  markerText,
  offsetPercent,
}: Readonly<ChildBmitGraphAxisMarkerProps>): JSX.Element {
  return (
    <div
      className="nhsuk-cbmi-axis-marker"
      style={{ left: `${offsetPercent}%` }}>
      <div className="nhsuk-font-size-16 nhsuk-cbmi-axis-marker-label">
        {markerText}
      </div>
    </div>
  );
}

export function ChildBmiChartAxes({
  model,
}: Readonly<ChildBmitGraphAxesProps>): JSX.Element {
  return (
    <>
      <AxisMarker
        markerText={model.bounds.upperBounds[
          BaseClassifications.Underweight
        ].toString()}
        offsetPercent={model.classificationPercentage(
          BaseClassifications.Underweight
        )}
      />
      <AxisMarker
        markerText={model.bounds.upperBounds[
          BaseClassifications.Healthy
        ].toString()}
        offsetPercent={model.classificationPercentage(
          BaseClassifications.Healthy
        )}
      />
    </>
  );
}
