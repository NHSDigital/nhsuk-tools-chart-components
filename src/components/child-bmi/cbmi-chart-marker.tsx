import { ChildBmiResultChart } from '@models/cbmi-chart.model';
import React, { JSX } from 'react';
import { CentileValue } from '@custom-types/bmi.types';
import { ChildBmiChartMarkerIcon } from './cbmi-chart-marker-icon';
import './styles/cbmi-chart-marker.module.scss';

/**
 * Function to determine which style to load for aligning the centile results text following below rules:
 * First 20% - alight left
 * Middle 60% - align center
 * Left 20% - alight right
 */
const alignTextStyle = (position: number): string => {
  if (position <= 20) {
    return 'nhsuk-cbmi-chart-label-align-left';
  }
  if (position <= 80) {
    return 'nhsuk-cbmi-chart-label-align-center';
  }
  return 'nhsuk-cbmi-chart-label-align-right';
};

function ChildBmiChartMarkerText({
  centile,
  markerText,
}: Readonly<{
  centile: CentileValue;
  markerText: string;
}>): JSX.Element {
  return (
    <div className="nhsuk-cbmi-chart-label-marker">
      <div className={`nhsuk-u-font-size-19 ${alignTextStyle(centile.value)}`}>
        {markerText}
        <span className="nhsuk-cbmi-chart-centile-marker">{centile.label}</span>
      </div>
    </div>
  );
}
export function ChildBmiChartMarker({
  centile,
  model,
  markerText,
}: Readonly<{
  centile: CentileValue;
  model: ChildBmiResultChart;
  markerText: string;
}>): JSX.Element {
  const offset = model.clampedOffsetPercentage(centile.value);

  let roundedOffset = 0;
  if (offset > 0) {
    const roundingPrecision = 1;
    const multiplier = 10 ** roundingPrecision;
    roundedOffset = Math.round(offset * multiplier) / multiplier;
  }

  return (
    <div>
      <div
        className="nhsuk-cbmi-chart-arrow-marker"
        style={{ left: `${roundedOffset}%` }}>
        <ChildBmiChartMarkerText centile={centile} markerText={markerText} />
        <ChildBmiChartMarkerIcon />
      </div>
      <div
        className="nhsuk-cbmi-chart-axis-marker"
        style={{ left: `${roundedOffset}%` }}
      />
    </div>
  );
}
