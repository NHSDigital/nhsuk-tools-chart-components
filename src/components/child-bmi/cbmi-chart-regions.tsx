import { ChildBmiResultChart } from '@models/cbmi-chart.model';
import {
  BaseClassificationLevel,
  BaseClassifications,
} from '@custom-types/bmi.types';
import React, { JSX } from 'react';
import './styles/cbmi-chart-regions.module.scss';

function RegionForClassification({
  classification,
  model,
}: Readonly<{
  classification: BaseClassificationLevel;
  model: ChildBmiResultChart;
}>): JSX.Element {
  const width = model.classificationPercentage(classification);
  return (
    <div
      className={`nhsuk-cbmi-classification-${classification}`}
      style={{ width: `${width}%` }}
    />
  );
}

export function ChildBmiChartRegions({
  model,
}: Readonly<{
  model: ChildBmiResultChart;
}>): JSX.Element {
  return (
    <>
      <RegionForClassification
        classification={BaseClassifications.Underweight}
        model={model}
      />
      <RegionForClassification
        classification={BaseClassifications.Healthy}
        model={model}
      />
      <RegionForClassification
        classification={BaseClassifications.Overweight}
        model={model}
      />
    </>
  );
}
