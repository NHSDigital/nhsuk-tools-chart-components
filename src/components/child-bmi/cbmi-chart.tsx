import { ChildBmiResultChart } from '@models/cbmi-chart.model';
import React, { JSX } from 'react';
import {
  BaseClassificationBounds,
  CentileValue,
} from '@custom-types/bmi.types';
import { ChildBmiChartAxes } from './cbmi-chart-axes';
import { ChildBmiChartMarker } from './cbmi-chart-marker';
import { ChildBmiChartRegions } from './cbmi-chart-regions';
import './styles/cbmi-chart.module.scss';

/**
 * The properties for @ChildBmiChart component
 *
 * @param {BaseClassificationBounds} classificationBounds -  The child body mass index classification ranges
 * @param {CentileValue} centile - The centile value and label
 * @param {string} markerText - Marker text for the centile value
 */
export type ChildBmiChartProps = {
  classificationBounds: BaseClassificationBounds;
  centile: CentileValue;
  markerText: string;
};

/**
 * A component to display the child body mass index chart
 *
 * @typedef {object} ChildBmiChartProps
 * @property {BaseClassificationBounds} classificationBounds -  The child body mass index classification ranges
 * @property {CentileValue} centile - The centile value and label
 * @property {string} markerText - Marker text for the centile value
 *
 * @param {ChildBmiChartProps} props
 * @returns {JSX.Element}
 */
export function ChildBmiChart({
  classificationBounds,
  centile,
  markerText,
}: Readonly<ChildBmiChartProps>): JSX.Element {
  const chartModel = new ChildBmiResultChart(classificationBounds);

  return (
    <div
      className={`nhsuk-cbmi-graph nhsuk-u-margin-bottom-9 `}
      aria-hidden="true">
      <ChildBmiChartAxes model={chartModel} />
      <ChildBmiChartMarker
        centile={centile}
        markerText={markerText}
        model={chartModel}
      />
      <ChildBmiChartRegions model={chartModel} />
    </div>
  );
}
