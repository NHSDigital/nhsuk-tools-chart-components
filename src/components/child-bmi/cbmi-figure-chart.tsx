import React, { JSX } from 'react';
import {
  BaseClassificationBounds,
  CentileValue,
  ChildBmiLegendKeys,
} from '@custom-types/bmi.types';
import { ChartLegend } from '@components/chart-legend';
import { ChildBmiChart } from './cbmi-chart';
import './styles/cbmi-chart-card.module.scss';

/**
 * The properties for @ChildBmiChartFigure component
 *
 * @param {BaseClassificationBounds} classificationBounds -  The child body mass index classification ranges
 * @param {CentileValue} centile - The centile value and label
 * @param {string} ariaLabel - The aria label for the chart figure
 * @param {string} centileMarkerText - The marker text for the centile value
 * @param {ChildBmiLegendKeys} legendKeys - The chart legend items
 */
export type ChildBmiChartFigureProps = {
  classificationBounds: BaseClassificationBounds;
  centile: CentileValue;
  ariaLabel: string;
  centileMarkerText: string;
  legendKeys: ChildBmiLegendKeys;
};

/**
 * A component to display the child body mass index chart figure
 *
 * @typedef {object} ChildBmiChartFigureProps
 * @property {BaseClassificationBounds} classificationBounds -  The child body mass index classification ranges
 * @property {CentileValue} centile - The centile value and label
 * @property {string} centileMarkerText - The marker text for the centile value
 * @property {string} ariaLabel - The aria label for the chart figure
 * @property {ChildBmiLegendKeys} legendKeys - The chart legend items
 *
 * @param {ChildBmiChartFigureProps} props
 * @returns {JSX.Element}
 */
export function ChildBmiChartFigure({
  classificationBounds,
  centile,
  centileMarkerText,
  ariaLabel,
  legendKeys,
}: Readonly<ChildBmiChartFigureProps>): JSX.Element {
  const cbmiLegendKeys = Object.entries(legendKeys).map(
    ([classification, label]) => ({
      text: label,
      className: `nhsuk-cbmi-legend-key-classification-${classification}`,
      wrapperClassName: 'nhsuk-cbmi-legend-marker',
    })
  );

  return (
    <figure className="nhsuk-u-margin-0">
      <figcaption className="nhsuk-u-visually-hidden">{ariaLabel}</figcaption>
      <ChildBmiChart
        classificationBounds={classificationBounds}
        centile={centile}
        markerText={centileMarkerText}
      />
      <ChartLegend legendKeys={cbmiLegendKeys} borderWrapper={false} />
    </figure>
  );
}
