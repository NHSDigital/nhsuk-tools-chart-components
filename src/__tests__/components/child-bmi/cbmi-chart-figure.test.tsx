import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

import React from 'react';
import {
  BaseClassificationBounds,
  CentileValue,
  ChildBmiLegendKeys,
} from '@custom-types/bmi.types';
import {
  mockedCentileValue,
  mockedChildBmiClassificationBounds,
} from '@tests/__helpers__/mocks/cbmi-results';
import { ChildBmiChartFigure } from '@components/child-bmi/cbmi-figure-chart';

const mockedAriaLabel = 'Child body mass index figure description';

const mockedLegendKeys: ChildBmiLegendKeys = {
  underweight: 'Underweight range',
  healthy: 'Healthy range',
  overweight: 'Overweight range',
};

function ChildBmiChartFigureTestApp({
  classificationBounds = mockedChildBmiClassificationBounds,
  centile = mockedCentileValue,
  markerText = '',
  ariaLabel = mockedAriaLabel,
  legendKeys = mockedLegendKeys,
}: Readonly<{
  classificationBounds?: BaseClassificationBounds;
  centile?: CentileValue;
  markerText?: string;
  ariaLabel?: string;
  legendKeys?: ChildBmiLegendKeys;
  borderWrapper?: boolean;
}>): JSX.Element {
  return (
    <ChildBmiChartFigure
      classificationBounds={classificationBounds}
      centile={centile}
      centileMarkerText={markerText}
      ariaLabel={ariaLabel}
      legendKeys={legendKeys}
    />
  );
}

describe('The Child body mass index chart figure component', () => {
  it('Renders with the expected content', () => {
    const { container } = render(<ChildBmiChartFigureTestApp />);
    expect(container).toMatchSnapshot();
  });

  it('Renders a figcaption with aria label', () => {
    const { container } = render(
      <ChildBmiChartFigureTestApp ariaLabel={mockedAriaLabel} />
    );
    const figcaptionEl = container?.querySelector('figcaption');

    expect(figcaptionEl).not.toBeNull();
    expect(figcaptionEl?.classList).toContain('nhsuk-u-visually-hidden');
    expect(figcaptionEl?.innerHTML).toBe(mockedAriaLabel);
  });

  it('Renders with the expected marker keys', () => {
    const { container } = render(<ChildBmiChartFigureTestApp />);

    // its borderless component, card should not be rendered
    const legendCardEl = container?.querySelector('.nhsuk-chart-legend-card');
    expect(legendCardEl).toBeNull();

    const legendKeyWrappersEl = container?.querySelectorAll(
      '.nhsuk-chart-legend-key-wrapper'
    );
    expect(legendKeyWrappersEl).not.toBeNull();
    expect(legendKeyWrappersEl?.length).toBe(3);

    Object.entries(mockedLegendKeys).forEach(([key, value], index) => {
      const legendKeyWrapper = legendKeyWrappersEl![index];
      const legendKeyEl = legendKeyWrapper?.querySelector(
        `.nhsuk-cbmi-legend-key-classification-${key}`
      );
      expect(legendKeyEl).not.toBeNull();
      expect(legendKeyWrapper.innerHTML).toContain(value);
    });
  });
});
