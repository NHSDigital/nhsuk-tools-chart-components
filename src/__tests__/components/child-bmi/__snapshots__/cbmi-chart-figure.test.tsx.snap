// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Child body mass index chart figure component Renders with the expected content 1`] = `
<div>
  <figure
    class="nhsuk-u-margin-0"
  >
    <figcaption
      class="nhsuk-u-visually-hidden"
    >
      Child body mass index figure description
    </figcaption>
    <div
      aria-hidden="true"
      class="nhsuk-cbmi-graph nhsuk-u-margin-bottom-9 "
    >
      <div
        class="nhsuk-cbmi-axis-marker"
        style="left: 8%;"
      >
        <div
          class="nhsuk-font-size-16 nhsuk-cbmi-axis-marker-label"
        >
          2
        </div>
      </div>
      <div
        class="nhsuk-cbmi-axis-marker"
        style="left: 80%;"
      >
        <div
          class="nhsuk-font-size-16 nhsuk-cbmi-axis-marker-label"
        >
          91
        </div>
      </div>
      <div>
        <div
          class="nhsuk-cbmi-chart-arrow-marker"
          style="left: 4%;"
        >
          <div
            class="nhsuk-cbmi-chart-label-marker"
          >
            <div
              class="nhsuk-u-font-size-19 nhsuk-cbmi-chart-label-align-left"
            >
              <span
                class="nhsuk-cbmi-chart-centile-marker"
              >
                Below 2
              </span>
            </div>
          </div>
          <svg
            fill="none"
            viewBox="0 0 12 12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 9L0.803849 -9.78799e-07L11.1962 -7.02746e-08L6 9Z"
              fill="black"
            />
          </svg>
        </div>
        <div
          class="nhsuk-cbmi-chart-axis-marker"
          style="left: 4%;"
        />
      </div>
      <div
        class="nhsuk-cbmi-classification-underweight"
        style="width: 8%;"
      />
      <div
        class="nhsuk-cbmi-classification-healthy"
        style="width: 80%;"
      />
      <div
        class="nhsuk-cbmi-classification-overweight"
        style="width: 100%;"
      />
    </div>
    <div
      aria-hidden="true"
    >
      <div
        class="nhsuk-u-margin-top-6 nhsuk-u-margin-bottom-3 nhsuk-u-secondary-text-color nhsuk-body-s"
      >
        <div
          class="nhsuk-chart-legend-key-wrapper nhsuk-cbmi-legend-marker"
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-cbmi-legend-key-classification-underweight"
          />
          Underweight range
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper nhsuk-cbmi-legend-marker"
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-cbmi-legend-key-classification-healthy"
          />
          Healthy range
        </div>
        <div
          class="nhsuk-chart-legend-key-wrapper nhsuk-cbmi-legend-marker"
        >
          <div
            class="nhsuk-chart-legend-key nhsuk-cbmi-legend-key-classification-overweight"
          />
          Overweight range
        </div>
      </div>
    </div>
  </figure>
</div>
`;
